### PR TITLE
Tf0.12 pipeline

### PR DESCRIPTION
**Pull request by jamesgorrie**
Upgrade to tf0.12 for 💸 & 🎉 

Thought about moving the AWS account as well, might do once this is done.

So far I have the sierra transformer and supporting infra, will keep ploughing through.

Related https://github.com/wellcometrust/platform/issues/4257

Supersedes #412

### CODE REVIEW

File: pipeline/terraform/locals.tf [hunk 1]
```diff
@@ -11,7 +11,7 @@ locals {
   # Sierra adapter topics
   sierra_merged_items_topic_name = "${data.terraform_remote_state.sierra_adapter.merged_items_topic_name}"
   sierra_merged_bibs_topic_name  = "${data.terraform_remote_state.sierra_adapter.merged_bibs_topic_name}"
-
+
```
Comment (alexwlchan) [inline pipeline/terraform/locals.tf]: boooo why is terraform formatting broken :(

File: pipeline/terraform0.12/modules/pipeline_service/variables.tf [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+variable "launch_type" {
+  default = "FARGATE"
+}
+
+variable "cpu" {
+  default = "256"
+}
+
+variable "memory" {
+  default = "512"
```
Comment (alexwlchan) [inline pipeline/terraform0.12/modules/pipeline_service/variables.tf]: type hints?

File: pipeline/terraform0.12/terraform.tf [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+terraform {
+  required_version = ">= 0.9"
+
+  backend "s3" {
+    role_arn = "arn:aws:iam::760097843905:role/platform-developer"
+
+    bucket         = "wellcomecollection-platform-infra"
+    key            = "terraform/catalogue_pipeline_tf_0.12.tfstate"
```
Comment (alexwlchan) [inline pipeline/terraform0.12/terraform.tf]:
```suggestion
    key            = "terraform/catalogue/pipeline.tfstate"
```

I've been trying to sort terraform states into per-repo folders in the S3 bucket, so it's a bit easier to find where any given state file is defined from. The platform-infra bucket in particular has a lot of state files now.

Also, let's not bake the terraform version into the name – that feels like a naming smell.

File: pipeline/terraform0.12/terraform.tf [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    dynamodb_table = "terraform-locktable"
+    region         = "eu-west-1"
+  }
+}
+
+data "terraform_remote_state" "catalogue_infra_critical" {
+  backend = "s3"
+
+  config = {
+    role_arn = "arn:aws:iam::760097843905:role/platform-developer"
```
Comment (alexwlchan) [inline pipeline/terraform0.12/terraform.tf]: Anything reading remote state should use a read-only role.

File: pipeline/terraform0.12/modules/pipeline_service/variables.tf [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+variable "aws_region" {}
+variable "namespace_id" {}
+variable "container_image" {}
+
+variable "secret_env_vars" {
+  type = map(string)
+}
+
+variable "secret_env_vars_length" {}
```
Comment (alexwlchan) [inline pipeline/terraform0.12/modules/pipeline_service/variables.tf]: All env_vars_length variables should be purged! Purged!

(This is one of the very nice things about terraform 0.12 that we get to ditch.)

File: pipeline/terraform0.12/terraform.tf (hunk 3 above)
Comment (kenoir) [reply pipeline/terraform0.12/terraform.tf]: I also think this will not work as you need the state from the stack pre update.

**Pull request closed without merging**